File: app/src/main/java/com/mohnage7/movies/view/adapter/VideoAdapter.java
package com.mohnage7.movies.view.adapter;

import android.app.Activity;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.youtube.player.YouTubeInitializationResult;
import com.google.android.youtube.player.YouTubeStandalonePlayer;
import com.google.android.youtube.player.YouTubeThumbnailLoader;
import com.google.android.youtube.player.YouTubeThumbnailView;
import com.mohnage7.movies.R;
import com.mohnage7.movies.base.BaseViewHolder;
import com.mohnage7.movies.model.Video;
import com.mohnage7.movies.utils.Constants;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class VideoAdapter extends RecyclerView.Adapter<BaseViewHolder> {
    private List<Video> videoList;

    public VideoAdapter(List<Video> videoList) {
        this.videoList = videoList;
    }

    @NonNull
    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_video, parent, false);
        return new MoviesViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        holder.bindViews(position);
    }

    @Override
    public int getItemCount() {
        if (videoList != null && !videoList.isEmpty()) {
            return videoList.size();
        } else {
            return 0;
        }
    }


    protected class MoviesViewHolder extends BaseViewHolder {
        @BindView(R.id.youtube_thumbnail)
        YouTubeThumbnailView youTubeThumbnailView;
        @BindView(R.id.play_button)
        ImageButton playButton;
        @BindView(R.id.tailer_title_txt_view)
        TextView trailerTitleTxtView;

        MoviesViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void bindViews(int position) {
            super.bindViews(position);
            final Video video = videoList.get(position);
            // set video title
            trailerTitleTxtView.setText(video.getName());
            // load video thumbnail
            final YouTubeThumbnailLoader.OnThumbnailLoadedListener onThumbnailLoadedListener = new YouTubeThumbnailLoader.OnThumbnailLoadedListener() {
                @Override
                public void onThumbnailError(YouTubeThumbnailView youTubeThumbnailView, YouTubeThumbnailLoader.ErrorReason errorReason) {
                    playButton.setVisibility(View.GONE);
                }

                @Override
                public void onThumbnailLoaded(YouTubeThumbnailView youTubeThumbnailView, String s) {
                    youTubeThumbnailView.setVisibility(View.VISIBLE);
                    playButton.setVisibility(View.VISIBLE);
                }
            };
            // init player
            youTubeThumbnailView.initialize(Constants.YOUTUBE_KEY, new YouTubeThumbnailView.OnInitializedListener() {
                @Override
                public void onInitializationSuccess(YouTubeThumbnailView youTubeThumbnailView, YouTubeThumbnailLoader youTubeThumbnailLoader) {
                    youTubeThumbnailLoader.setVideo(video.getKey());
                    youTubeThumbnailLoader.setOnThumbnailLoadedListener(onThumbnailLoadedListener);
                 }

                @Override
                public void onInitializationFailure(YouTubeThumbnailView youTubeThumbnailView, YouTubeInitializationResult youTubeInitializationResult) {
                    //write something for failure
                }
            });

            // play video in full screen activity when user clicks
            playButton.setOnClickListener(v -> {
                Intent intent = YouTubeStandalonePlayer.createVideoIntent((Activity) itemView.getContext(), Constants.YOUTUBE_KEY, video.getKey());
                itemView.getContext().startActivity(intent);
            });
        }


        @Override
        public void clear() {
            playButton.setOnClickListener(null);
            trailerTitleTxtView.setText("");
        }
    }
}

File: app/src/main/java/com/mohnage7/movies/features/moviedetails/view/VideoAdapter.kt
package com.mohnage7.movies.features.moviedetails.view

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.android.youtube.player.YouTubeStandalonePlayer
import com.mohnage7.movies.BuildConfig.YOUTUBE_KEY
import com.mohnage7.movies.R
import com.mohnage7.movies.base.BaseViewHolder
import com.mohnage7.movies.features.moviedetails.model.Video
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_video.view.*


const val YOUTUBE_THUMBNAIL_BASE_URL = "https://img.youtube.com/vi/"
const val YOUTUBE_THUMBNAIL_URL_JPG = "/0.jpg"

class VideoAdapter(private val videoList: List<Video>?) : RecyclerView.Adapter<BaseViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_video, parent, false)
        return MoviesViewHolder(view)
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.bindViews(position)
    }

    override fun getItemCount(): Int {
        return if (videoList != null && !videoList.isEmpty()) {
            videoList.size
        } else {
            0
        }
    }


    private inner class MoviesViewHolder internal constructor(itemView: View) : BaseViewHolder(itemView) {

        override fun bindViews(position: Int) {
            super.bindViews(position)
            val video = videoList!![position]
            // set video title
            itemView.trailerTitleTxtView.text = video.name

            Picasso.get()
                    .load(video.getVideoThumbnail())
                    .into(itemView.youTubeThumbnailView, object : Callback {
                        override fun onError(e: Exception?) {
                            itemView.playButton.visibility = View.GONE
                        }

                        override fun onSuccess() {
                            itemView.playButton.visibility = View.VISIBLE
                        }
                    })

            // play video in full screen activity when user clicks
            itemView.playButton!!.setOnClickListener {
                val intent = YouTubeStandalonePlayer.createVideoIntent(itemView.context as Activity, YOUTUBE_KEY, video.key!!)
                itemView.context.startActivity(intent)
            }
        }


        override fun clear() {
            itemView.playButton.setOnClickListener(null)
            itemView.trailerTitleTxtView.text = ""
            itemView.youTubeThumbnailView.setImageDrawable(null)
        }
    }
}
